### PR TITLE
Fix pytype static checks

### PR DESCRIPTION
**Pull request by hjmjohnson**
### Description
Initial status for testing the use of pytype static checking.

### Status
Ready

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by hjmjohnson** [comment]
@wyli @Nic-Ma Would you mind reviewing this PR?  We have more work to do, but this demonstrates the direction that we are moving towards.

File: monai/data/csv_saver.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ class CSVSaver:
     the cached data into CSV file. If no meta data provided, use index from 0 to save data.
     """
 
-    def __init__(self, output_dir="./", filename="predictions.csv", overwrite=True):
+    def __init__(self, output_dir: str = "./", filename: str = "predictions.csv", overwrite: bool = True):
         """
         Args:
             output_dir (str): output CSV file directory.
```
Comment (hjmjohnson) [inline monai/data/csv_saver.py:30]: @benjamin-gorman @crtrentz You need to remove the types in the comments. 

Once you define `output_dir: str`  using PEP-484 you need to change `output_dir (str):` to `output_dir:`

**Comment by Nic-Ma** [comment]
Hi @hjmjohnson ,

Thanks for your update on the pytype check.
Overall looks good to me.
As @wyli spent time to help on this topic before, I think he can review more clearly.
Thanks.

**Comment by wyli** [comment]
thanks! I checked out this branch, running `./runtests.sh --codeformat --quick` 
with the settings:
```
(py36) ➜  flake8 --version
3.8.1 (flake8-bugbear: 20.1.4, flake8-comprehensions: 3.2.2,
flake8-executable: 2.0.3, flake8-mypy: 17.8.0, flake8-pyi: 20.5.0, mccabe:
0.6.1, naming: 0.10.0, pycodestyle: 2.6.0, pyflakes: 2.2.0) CPython 3.6.10 on
Darwin
```
and it gives the following error:
```
*** black code format tests passed. ***
-----------------------------------
Verifying flake8 formatting checks.
/Users/wenqili/Documents/MONAI/monai/losses/dice.py:67:6: T484 Signature of "forward" incompatible with supertype "Module"
/Users/wenqili/Documents/MONAI/monai/losses/dice.py:113:20: T484 Unsupported operand types for - ("float" and "Tensor")
/Users/wenqili/Documents/MONAI/monai/losses/dice.py:115:21: T484 "float" has no attribute "sum"
/Users/wenqili/Documents/MONAI/monai/losses/dice.py:119:21: T484 "float" has no attribute "mean"
/Users/wenqili/Documents/MONAI/monai/losses/dice.py:170:6: T484 Signature of "forward" incompatible with supertype "Module"
/Users/wenqili/Documents/MONAI/monai/losses/dice.py:211:21: T484 Module has no attribute "isinf"
/Users/wenqili/Documents/MONAI/monai/losses/tversky.py:71:6: T484 Signature of "forward" incompatible with supertype "Module"
/Users/wenqili/Documents/MONAI/monai/losses/tversky.py:116:24: T484 Unsupported operand types for - ("float" and "Tensor")
/Users/wenqili/Documents/MONAI/monai/losses/tversky.py:119:21: T484 "float" has no attribute "sum"
/Users/wenqili/Documents/MONAI/monai/losses/tversky.py:123:21: T484 "float" has no attribute "mean"
/Users/wenqili/Documents/MONAI/monai/transforms/spatial/array.py:884:16: T484 Unexpected keyword argument "align_corners" for "grid_sample"
/Users/wenqili/Documents/MONAI/monai/data/grid_dataset.py:16:2: T484 Module 'torch.utils.data' has no attribute 'IterableDataset'
12    T484 Module 'torch.utils.data' has no attribute 'IterableDataset'
12
-----------------------------------
----------------------------------
Formatting test failed!
```
any idea?

**Comment by wyli** [comment]
also, MONAI uses a "squash and merge" method for pull requests, which means we'll lose some co-author information in the master branch when we merge this PR (e.g. `git blame` command won't work accurately). is that fine with you?

otherwise I'm happy to review changes on this topic in multiple, incremental PRs from each author as well...

**Comment by hjmjohnson** [comment]
Yes.   That is mypy flake inconsistency.   We will address all of those.   Would you mind commenting if you have any concerns with the overall approach?  We are planning to spend more than a week of effort making these static checkers consistent,   But we want feedback to ensure that it is not wasted effort.

**Comment by wyli** [comment]
yes, adding type hints looks good to me. @ericspod might have further comments as well?

**Comment by hjmjohnson** [comment]
Great.  We will press forward then.

**Comment by hjmjohnson** [comment]
> thanks! I checked out this branch, running `./runtests.sh --codeformat --quick`
> with the settings:
> 
> ```
> (py36) ➜  flake8 --version
> 3.8.1 (flake8-bugbear: 20.1.4, flake8-comprehensions: 3.2.2,
> flake8-executable: 2.0.3, flake8-mypy: 17.8.0, flake8-pyi: 20.5.0, mccabe:
> 0.6.1, naming: 0.10.0, pycodestyle: 2.6.0, pyflakes: 2.2.0) CPython 3.6.10 on
> Darwin
> ```
> 
> and it gives the following error:
> 
> ```
> *** black code format tests passed. ***
> -----------------------------------
> Verifying flake8 formatting checks.
> /Users/wenqili/Documents/MONAI/monai/losses/dice.py:67:6: T484 Signature of "forward" incompatible with supertype "Module"
> /Users/wenqili/Documents/MONAI/monai/losses/dice.py:113:20: T484 Unsupported operand types for - ("float" and "Tensor")
> /Users/wenqili/Documents/MONAI/monai/losses/dice.py:115:21: T484 "float" has no attribute "sum"
> /Users/wenqili/Documents/MONAI/monai/losses/dice.py:119:21: T484 "float" has no attribute "mean"
> /Users/wenqili/Documents/MONAI/monai/losses/dice.py:170:6: T484 Signature of "forward" incompatible with supertype "Module"
> /Users/wenqili/Documents/MONAI/monai/losses/dice.py:211:21: T484 Module has no attribute "isinf"
> /Users/wenqili/Documents/MONAI/monai/losses/tversky.py:71:6: T484 Signature of "forward" incompatible with supertype "Module"
> /Users/wenqili/Documents/MONAI/monai/losses/tversky.py:116:24: T484 Unsupported operand types for - ("float" and "Tensor")
> /Users/wenqili/Documents/MONAI/monai/losses/tversky.py:119:21: T484 "float" has no attribute "sum"
> /Users/wenqili/Documents/MONAI/monai/losses/tversky.py:123:21: T484 "float" has no attribute "mean"
> /Users/wenqili/Documents/MONAI/monai/transforms/spatial/array.py:884:16: T484 Unexpected keyword argument "align_corners" for "grid_sample"
> /Users/wenqili/Documents/MONAI/monai/data/grid_dataset.py:16:2: T484 Module 'torch.utils.data' has no attribute 'IterableDataset'
> 12    T484 Module 'torch.utils.data' has no attribute 'IterableDataset'
> 12
> -----------------------------------
> ----------------------------------
> Formatting test failed!
> ```
> 
> any idea?

@wyli  I am unable to produce the warnings now on either mac or linux environments.  Perhaps it was a problem with the previous codebase that we have fixed.  

Would you mind trying again?

Thanks,
Hans

**Comment by hjmjohnson** [comment]
@Nic-Ma @wyli @ericspod  I realize that this set of changes is incomplete, but they represent a self-contained step towards the eventual goal of supporting typing for MONAI.

Would you please consider if this could be the first step towards a much more complete solution?

Thanks,
Hans

**Comment by ericspod** [comment]
Moving forward with using the optional is a really good idea. Ignite is doing this too so this keeps us aligned with them and others starting to use the annotations. Flake8 is using mypy and not pytype correct? The error output [here](https://github.com/Project-MONAI/MONAI/pull/449#issuecomment-636351235) mentions mypy. Pytype is more lenient and does type inference for unannotated values, I'd think this would be what we'd want. There is also pyre-check but that looks a bit more complicated with a running server concept. We should move forward with this but all developers should be made aware that using annotations is expected now and information on code style needs changing. Are we thinking to merge this now or wait for complete annotations?

**Comment by hjmjohnson** [comment]
@ericspod I think that waiting for complete annotations would be a mistake.  This is not a trivial task, and the potential for code conflicts is quite high.  I have a team of students working through adding code annotations for the next few weeks, but if they wait until the end to merge we will have a huge burden of managing code conflicts with every other PR.    

I propose a commit early and commit often approach over the next month to integrate the annotations.

**Comment by wyli** [comment]
> > thanks! I checked out this branch, running `./runtests.sh --codeformat --quick`
> > Formatting test failed!
> > ```
> > 
> > 
> > any idea?
> 
> @wyli I am unable to produce the warnings now on either mac or linux environments. Perhaps it was a problem with the previous codebase that we have fixed.
> 
> Would you mind trying again?
> 
> Thanks,
> Hans

Hi @hjmjohnson the warnings come from pytorch 1.4, upgrading to 1.5.0 resolves the issue.
I'll merge this PR for now, and we may need to update the requirements if there are further CI issues:
https://github.com/Project-MONAI/MONAI/blob/0e5c3ae21028f6281861fee621bc44a183e5208e/requirements.txt#L1

**Pull request merged**